src/services/dynamodb.rs

```diff
@@ -34,3 +42,79 @@ async fn dynamodb_client(region: Option<&'static str>) -> DynamoDBClient {
 pub async fn dynamodb(region: Option<&'static str>) -> DynamoDBClient {
     dynamodb_client(region).await
 }
+
+#[derive(Serialize, Deserialize)]
+#[serde(transparent)]
+pub struct Version(u64);
+
+impl Version {
+    pub fn inc(&self) -> Self {
+        Version(self.0 + 1)
+    }
+}
+
+impl Display for Version {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+pub trait DynamoDBModel: Serialize + DeserializeOwned {}
+
+pub struct VersionField {
+    name: &'static str,
+    value: Version,
+}
+
+pub trait Versioned: DynamoDBModel {
+    fn version_field(&self) -> VersionField;
+}
+
+#[derive(Error, Debug)]
+pub enum VersionedPutError {
+    #[error(
+        "The item version mismatched the one in the database, so operation was not successful. This error is retryable after fetching a new version of the item"
+    )]
+    VersionMismatch,
+    #[error("Serialization error: {0}")]
+    SerializationError(#[from] serde_dynamo::Error),
+    #[error("DynamoDB error: {0}")]
+    OtherDynamoError(#[from] aws_sdk_dynamodb::Error),
+}
+
+/// Performs a DynamoDB Put operation with the provided item.
+/// The operation is versioned, meaning it will fail if the version
+/// in the database mismatches the one in the provided item.
+pub async fn versioned_put<T>(
```

Comment:
Is there some Rust magic where we can make it generic over Put and Update? Like a version-"flavored" client or update/put operation.

Reply:
I don't think we want put make it generic. Put (create/replace) and update (update specific fields) are inherently different operations, so merging them seems wrong. But we can do a versioned_update in another pr.